Improve theme.json test failure messages by pretty printing css for a more accurate diff

talldan opened the pull request:
## What?
Related #60842

One of the issues with the theme json tests (the ones that people hate) is that when they fail they output the css diff as a big one-line string, and it's pretty much impossible to spot the difference. I had an idea to improve this, which is to apply some basic pretty printing to the CSS used in the assertion so that it's like a traditional stylesheet - on multiple lines.

When a test fails it'll now show a much more useful diff with only the specific selectors or rules that are actually different rather than all the css. I think this will make the tests much more useful, when making changes to production you'll now be able to see the exact impact.

This has been achieved using a custom assertion function that pretty prints the css strings.

The raw CSS is still useful for updating the failed test. To handle this, a custom message is appended on to the test failure containing the raw css. It does this by taking advantage of the fact that PHP unit test failures are thrown as exceptions. The exception from an assertion can be caught and re-thrown but with some extra information appended.

## Testing Instructions
1. Introduce an error in the expected css in one of the updates tests (preferably one of the ones with a big blog of css)
2. Run the tests
3. Observe that the diff output is now much more specific

## Screenshots or screencast <!-- if applicable -->
#### Before
```
There was 1 failure:

1) WP_Theme_JSON_Gutenberg_Test::test_get_stylesheet_generates_layout_styles_with_spacing_presets
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-':root { --wp--style--global--content-size: 640px;--wp--style--global--wide-size: 1200px; }:where(body) { margin: 0; }.wp-site-blocks > .alignleft { float: left; margin-right: 2em; }.wp-site-blocks > .alignright { float: right; margin-left: 2em; }.wp-site-blocks > .aligncenter { justify-content: center; margin-left: auto; margin-right: auto; }:where(.wp-site-blocks) > * { margin-block-start: var(--wp--preset--spacing--60); margin-block-end: 0; }:where(.wp-site-blocks) > :first-child { margin-block-start: 0; }:where(.wp-site-blocks) > :last-child { margin-block-end: 0; }:root { --wp--style--block-gap: var(--wp--preset--spacing--60); }.is-layout-flow  > :first-child{margin-block-start: 0;}.is-layout-flow  > :last-child{margin-block-end: 0;}.is-layout-flow  > *{margin-block-start: var(--wp--preset--spacing--60);margin-block-end: 0;}.is-layout-constrained  > :first-child{margin-block-start: 0;}.is-layout-constrained  > :last-child{margin-block-end: 0;}.is-layout-constrained  > *{margin-block-start: var(--wp--preset--spacing--60);margin-block-end: 0;}.is-layout-flex {gap: var(--wp--preset--spacing--60);}.is-layout-grid {gap: var(--wp--preset--spacing--60);}.is-layout-flow > .alignleft{float: left;margin-inline-start: 0;margin-inline-end: 2em;}.is-layout-flow > .alignright{float: right;margin-inline-start: 2em;margin-inline-end: 0;}.is-layout-flow > .aligncenter{margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > .alignleft{float: left;margin-inline-start: 0;margin-inline-end: 2em;}.is-layout-constrained .alignright{float: right;margin-inline-start: 2em;margin-inline-end: 0;}.is-layout-constrained > .aligncenter{margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > :where(:not(.alignleft):not(.alignright):not(.alignfull)){max-width: var(--wp--style--global--content-size);margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > .alignwide{max-width: var(--wp--style--global--wide-size);}body .is-layout-flex{display: flex;}.is-layout-flex{flex-wrap: wrap;align-items: center;}.is-layout-flex > :is(*, div){margin: 0;}body .is-layout-grid{display: grid;}.is-layout-grid > :is(*, div){margin: 0;}'
+':root { --wp--style--global--content-size: 640px;--wp--style--global--wide-size: 1200px; }:where(body) { margin: 0; }.wp-site-blocks > .alignleft { float: left; margin-right: 2em; }.wp-site-blocks > .alignright { float: right; margin-left: 2em; }.wp-site-blocks > .aligncenter { justify-content: center; margin-left: auto; margin-right: auto; }:where(.wp-site-blocks) > * { margin-block-start: var(--wp--preset--spacing--60); margin-block-end: 0; }:where(.wp-site-blocks) > :first-child { margin-block-start: 0; }:where(.wp-site-blocks) > :last-child { margin-block-end: 0; }:root { --wp--style--block-gap: var(--wp--preset--spacing--60); }.is-layout-flow  > :first-child{margin-block-start: 0;}.is-layout-flow  > :last-child{margin-block-end: 0;}.is-layout-flow  > *{margin-block-start: var(--wp--preset--spacing--60);margin-block-end: 0;}.is-layout-constrained  > :first-child{margin-block-start: 0;}.is-layout-constrained  > :last-child{margin-block-end: 0;}.is-layout-constrained  > *{margin-block-start: var(--wp--preset--spacing--60);margin-block-end: 0;}.is-layout-flex {gap: var(--wp--preset--spacing--60);}.is-layout-grid {gap: var(--wp--preset--spacing--60);}.is-layout-flow > .alignleft{float: left;margin-inline-start: 0;margin-inline-end: 2em;}.is-layout-flow > .alignright{float: right;margin-inline-start: 2em;margin-inline-end: 0;}.is-layout-flow > .aligncenter{margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > .alignleft{float: left;margin-inline-start: 0;margin-inline-end: 2em;}.is-layout-constrained > .alignright{float: right;margin-inline-start: 2em;margin-inline-end: 0;}.is-layout-constrained > .aligncenter{margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > :where(:not(.alignleft):not(.alignright):not(.alignfull)){max-width: var(--wp--style--global--content-size);margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > .alignwide{max-width: var(--wp--style--global--wide-size);}body .is-layout-flex{display: flex;}.is-layout-flex{flex-wrap: wrap;align-items: center;}.is-layout-flex > :is(*, div){margin: 0;}body .is-layout-grid{display: grid;}.is-layout-grid > :is(*, div){margin: 0;}'

/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:1122
```


#### After
```
2) WP_Theme_JSON_Gutenberg_Test::test_get_stylesheet_generates_layout_styles_with_spacing_presets
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
 margin-inline-end: 2em;
 }

-.is-layout-constrained .alignright {
+.is-layout-constrained > .alignright {
 float: right;
 margin-inline-start: 2em;
 margin-inline-end: 0;

/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:77
/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:1157


If the change is expected, update the test case to this CSS:
:root { --wp--style--global--content-size: 640px;--wp--style--global--wide-size: 1200px; }:where(body) { margin: 0; }.wp-site-blocks > .alignleft { float: left; margin-right: 2em; }.wp-site-blocks > .alignright { float: right; margin-left: 2em; }.wp-site-blocks > .aligncenter { justify-content: center; margin-left: auto; margin-right: auto; }:where(.wp-site-blocks) > * { margin-block-start: var(--wp--preset--spacing--60); margin-block-end: 0; }:where(.wp-site-blocks) > :first-child { margin-block-start: 0; }:where(.wp-site-blocks) > :last-child { margin-block-end: 0; }:root { --wp--style--block-gap: var(--wp--preset--spacing--60); }.is-layout-flow  > :first-child{margin-block-start: 0;}.is-layout-flow  > :last-child{margin-block-end: 0;}.is-layout-flow  > *{margin-block-start: var(--wp--preset--spacing--60);margin-block-end: 0;}.is-layout-constrained  > :first-child{margin-block-start: 0;}.is-layout-constrained  > :last-child{margin-block-end: 0;}.is-layout-constrained  > *{margin-block-start: var(--wp--preset--spacing--60);margin-block-end: 0;}.is-layout-flex {gap: var(--wp--preset--spacing--60);}.is-layout-grid {gap: var(--wp--preset--spacing--60);}.is-layout-flow > .alignleft{float: left;margin-inline-start: 0;margin-inline-end: 2em;}.is-layout-flow > .alignright{float: right;margin-inline-start: 2em;margin-inline-end: 0;}.is-layout-flow > .aligncenter{margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > .alignleft{float: left;margin-inline-start: 0;margin-inline-end: 2em;}.is-layout-constrained > .alignright{float: right;margin-inline-start: 2em;margin-inline-end: 0;}.is-layout-constrained > .aligncenter{margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > :where(:not(.alignleft):not(.alignright):not(.alignfull)){max-width: var(--wp--style--global--content-size);margin-left: auto !important;margin-right: auto !important;}.is-layout-constrained > .alignwide{max-width: var(--wp--style--global--wide-size);}body .is-layout-flex{display: flex;}.is-layout-flex{flex-wrap: wrap;align-items: center;}.is-layout-flex > :is(*, div){margin: 0;}body .is-layout-grid{display: grid;}.is-layout-grid > :is(*, div){margin: 0;}

/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:84
/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:1157
```

talldan committed:
Add basic formatting function and apply to one test

WIP

Add comment

andrewserong commented:
Overall I love the idea of making these tests more readable and useful, by highlighting the actual CSS that has changed, I think that will help a lot in the developer experience of working with these 👍

>  In particular the 'expected' CSS in the tests is still technically one big single line string, so when updating it you can't simply copy/paste the test output, it has to be updated bit by bit. Though this might be a good thing, as I think it means less chance of missing something in the diff. 

In my experience with these tests, I think this could be a negative. It can be quite time consuming to update the tests, and ideally it would be possible to copy + paste the actual string in the error message to quickly update a test.

Would it be possible to try including the original received raw string in the third param of the assertion, to see if that can give folks a usable string to copy/paste to update the tests?

From my perspective in updating these tests, we want to try to achieve two things at once:

* Improve readability and usefulness of test errors
* Improve the ability to quickly update the test

talldan committed:
Add assertSameCSS expectation which outputs pretty printed test failure along with raw css for updating the test

Update accidentally committed test failure

Fix expected,actual order and minimize diff

Fix - make sure the actual is output in addition to the pretty printed failure, not the expected

Update test expectation so test passes

Revert "Update test expectation so test passes"

This reverts commit c9f655ac55aeee96badd9bbd14ee3bec3dce127d.

Update test expectation so test passes

talldan commented:
> From my perspective in updating these tests, we want to try to achieve two things at once, if we can:
> 
> - Improve readability and usefulness of test errors
> - Improve the ability to quickly update the test

Ok, I think I've found a way in [2a3e034](https://github.com/WordPress/gutenberg/pull/64077/commits/2a3e03425dc25a743ff0191bf09180b332a924d0) to have both the pretty printed failures and the raw css string for updating the test case.

It'll now output something like this:
```
There was 1 failure:

1) WP_Theme_JSON_Gutenberg_Test::test_get_stylesheet_preset_classes_work_with_compounded_selectors
Failed asserting that two strings are identical.
--- Expected
+++ Actual
@@ @@
-'.wp-block-heading.has-white-color {
+'.wp-bloc-heading.has-white-color {
 color: var(--wp--preset--color--white) !important;
 }

/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:77
/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:755


If the change is expected, update the test case to this CSS:
.wp-bloc-heading.has-white-color{color: var(--wp--preset--color--white) !important;}.wp-block-heading.has-white-background-color{background-color: var(--wp--preset--color--white) !important;}.wp-block-heading.has-white-border-color{border-color: var(--wp--preset--color--white) !important;}

/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:85
/var/www/html/wp-content/plugins/gutenberg/phpunit/class-wp-theme-json-test.php:755
```

The only downside is that it outputs the line of code for the custom assertion at the bottom as well as the actual line of code of the failure. Also the code might be considered sketchy/hacky, but I think it's ok.

andrewserong commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -556,10 +597,10 @@ public function test_get_stylesheet() {
-		$this->assertSame( $styles, $theme_json->get_stylesheet( array( 'styles' ) ) );
-		$this->assertSame( $presets, $theme_json->get_stylesheet( array( 'presets' ) ) );
-		$this->assertSame( $all, $theme_json->get_stylesheet() );
+		$this->assertSameCSS( $variables, $theme_json->get_stylesheet( array( 'variables' ) ) );
+		$this->assertSameCSS( $styles, $theme_json->get_stylesheet( array( 'styles' ) ) );
+		$this->assertSameCSS( $presets, $theme_json->get_stylesheet( array( 'presets' ) ) );
+		$this->assertSameCSS( $all, $theme_json->get_stylesheet() );
```
Very nice, I love how this abstraction means that we don't add any lines to individual tests 👍

andrewserong commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -46,6 +46,47 @@ public static function set_up_before_class() {
 
+	// Pretty print CSS in test assertions so that it provides a better diff when a test fails.
+	// Without this: the failing test outputs the entire css string as being different.
+	// With this: the failing test only highlights the specific CSS rule that is different.
```
Tiny formatting nit: should this be a doc comment instead?

andrewserong reviewed (approved):
Great work @talldan! The pretty printing in test failures is going to make a huge difference to anyone working on updating these tests, and the new `assertSameCSS` method still allows for quick copy/pasting to update the affected tests. Also the inclusion of the error message is a win as it helps nudge folks toward how to fix a failing test when the changes are expected 👍 

Looking good while testing locally running `npm run test:unit:php:base -- --filter WP_Theme_JSON_Gutenberg_Test`:

<img width="1092" alt="image" src="https://github.com/user-attachments/assets/aef4e146-4943-4ac0-b357-563b0abe731f">

✅ Changes in CSS are now pretty printed correctly
✅ Copy + pasting the expected string into the test gets the test passing again

> The only downside is that it outputs the line of code for the custom assertion at the bottom (85) as well as the actual line of code of the failure (755). Also the code might be considered sketchy/hacky, but I think it's ok.

The double output of lines here seems quite minor to me, and at least it flags where the error message is generated. Also, I see what you mean in terms of the code possibly being seen as hacky, but given the clear developer experience wins and that it'll help debug failing tests, I think it's worth it 👍

LGTM! 🚀

ramonjd commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -46,6 +46,47 @@ public static function set_up_before_class() {
+	 * @param string $expected The expected CSS.
+	 * @param string $actual   The actual CSS.
+	 */
+	private function assertSameCSS( $expected, $actual, $message = '' ) {
```
Great idea! I like!

ramonjd commented on phpunit/class-wp-theme-json-test.php:
```diff
@@ -46,6 +46,47 @@ public static function set_up_before_class() {
+	 */
+	private function assertSameCSS( $expected, $actual, $message = '' ) {
+		$pretty_expected = self::pretty_print_css( $expected );
+		$pretty_actual   = self::pretty_print_css( $actual );
```
> but outputs pretty printed test failures along with the raw css

Just checking - is it just that we want to prettify the failure output? 

If so, does it make sense to only prettify things in the `catch` block?

talldan replied:
> Just checking - is it just that we want to prettify the failure output?

It's more that the `assertSame` method produces a better diff if it's passed css that has line breaks rather than one long string. The idea is that it'll just show the individual selectors/rules that have changed rather than the whole of the css stylesheet as it does in trunk. Background is I'm working on a PR (https://github.com/WordPress/gutenberg/pull/64076), but with the way it works in trunk it's really difficult to know exactly which selectors I've updated with the change. Especially if you're not as familiar with theme json like me. 😄 

I don't think it'd be possible to get the same level of output by doing the pretty printing in the catch block, unless a custom diff function is used, but it feels like overkill.

talldan replied:
Examples here:
- Before: https://github.com/WordPress/gutenberg/actions/runs/10156649825/job/28085502142?pr=64076
- After: https://github.com/WordPress/gutenberg/actions/runs/10174137732/job/28139475922?pr=64076

ramonjd replied:
Ah, got it. Thanks for the explanation - I didn't think about the change comparison. 

As you were!

talldan committed:
Convert comment to docblock

talldan commented:
Thanks for the reviews, I'll merge this and then follow up with the same change in core.
